quickwit-search/src/root.rs

```diff
@@ -110,21 +113,24 @@ pub struct ErrorRetries {
 /// - Is the node ok, but just the split failing?
 /// - Did all requests fail? (Should we retry in that case?)
 fn analyze_errors(search_result: &SearchResultsByAddr) -> Option<ErrorRetries> {
-    let contains_errors = search_result.values().any(|res| {
-        res.is_err()
-            || res
-                .as_ref()
-                .map(|ok_res| ok_res.failed_requests.is_empty())
-                .unwrap_or(false)
+    let contains_retryable_errors = search_result.values().any(|res| {
+        res.as_ref().map_or(true, |ok_res| {
+            ok_res
+                .failed_splits
+                .iter()
+                .filter(|err| err.retryable_error)
+                .count()
+                > 0
+        })
     });
-    if !contains_errors {
+    if !contains_retryable_errors {
         return None;
     }
     let complete_failure_nodes: Vec<_> = search_result
```

Comment:
I suspect a single procedural loop would actually have been easier to understand here.

```rust 
let mut complete_failure_nodes = Vec::new();
let mut partial_or_no_failure_nodes = Vec::new();
for (addr, leaf_search_result) in search_result {
    match leaf_search_result {
        Ok(res) => {
           // ...
        }
        Err(err) => {
           complete_failure_nodes.push(addr);
        }
    }
}
```
The functional approach you used here makes it very hard to proofread that this is a effectively partitioning the nodes.
Of course, you can also use the partition method and add a helper function called `is_complete_failure` or something like that.

Reply:
I think the functional approach with the helper function is the most readable